File: python/src/asqav/verifier/oracle/adapters/asqav_native.py
"""Asqav-native adapter - the existing verify_receipt logic behind the seam.

Two real Asqav wire shapes route here, kept mutually exclusive at detection:

  - COMPLIANCE mode: the 3-key ``{payload, signature, anchors}`` envelope whose
    ``payload`` carries ``previousReceiptHash`` / ``issuer_id`` (the protectmcp:*
    shape). ML-DSA-65 (or Ed25519 for the test vectors) signs the canonical bytes
    of ``payload`` DIRECTLY (no pre-hash, no field strip); the chain hash is the
    SHA-256 of the predecessor payload's canonical bytes.
  - HASH mode: the default ``/sign`` output - a FLAT receipt with ``mode:"hash"``,
    ``payload:null``, a ``signature_b64`` (or ``signature``), and a ``hash`` over
    the action bytes. The signing input is the flat 11-field object the cloud
    rebuilds, not the receipt itself; ``_hash_mode_signing_input`` reconstructs it.

Scope: this adapter verifies the issuer signature, the hash-chain link, and
structural presence of the required fields. It does NOT check anchor liveness or
issued_at clock skew - the standalone ``verify_receipt`` carries those axes.

Key resolution and the compliance-mode structure check delegate to
``verify_receipt`` so the two surfaces stay byte-for-byte identical; the oracle
does not re-implement them. Canonicalisation goes through ``asqav_jcs``, asserted
byte-identical to ``verify_receipt.canonical_json`` by the cloud-parity test.
"""
from __future__ import annotations

from typing import Any

from asqav.verifier import verify_receipt as _vr

from ..adapter import ChainStep, FormatAdapter, SignatureMaterial
from ..canonical import asqav_jcs
from ..core import sha256_hex
from .acta import _is_lower_hex

#: Field set the cloud's hash-mode signer canonicalises, in _build_signing_message order.
_HASH_MODE_FIELDS = (
    "v",
    "mode",
    "hash",
    "hash_algo",
    "metadata",
    "server_timestamp",
    "action_id",
    "agent_id",
    "org_id",
    "policy_digest",
    "policy_decision",
)


#: Claims belonging to the signed payload; hash mode signs the flat fields only,
#: so a thumbprint pasted onto one binds nothing.
_UNSIGNED_CLAIM_FIELDS = ("issuer_id", "previousReceiptHash", "key_thumbprint")


def _is_hash_mode(doc: dict) -> bool:
    """True for a flat hash-mode signature receipt (mode=hash, null payload, a sig).

    Routing reads the shape of the signed unit and nothing else: a dict
    ``payload`` is the compliance signed unit, any other shape routes here. No
    field an attacker can add or omit moves a doc between the paths, so an added
    field cannot select a weaker axis set. A doc displaying claims outside its
    signed unit FAILs the structure axis instead of being rerouted.
    """
    if doc.get("mode") != "hash" or isinstance(doc.get("payload"), dict):
        return False
    return bool(doc.get("signature_b64") or doc.get("signature"))


def _resolved_key_material(entry: dict | None) -> tuple[Any, bytes | None]:
    """alg plus raw public-key bytes of the resolved directory entry.

    The directory publishes the key as standard base64 under ``public_key``; an
    AKP thumbprint is taken over the same bytes in unpadded base64url, so the
    decode has to happen before the digest, never after.
    """
    if not isinstance(entry, dict):
        return None, None
    try:
        return entry.get("alg"), _vr._b64decode(entry.get("public_key", ""))
    except Exception:
        return entry.get("alg"), None


def _payload(doc: dict) -> dict:
    """Normalise to the signed payload regardless of envelope nesting.

    An explicit ``payload: null`` falls back to the doc, matching the TypeScript
    payloadOf, so a caller never has to guard against None.
    """
    env = _vr.normalise_envelope(doc)
    payload = env.get("payload")
    return payload if isinstance(payload, dict) else env


    # Decode signature material; b'' on any malformed input so verify FAILs, never crashes.
def _safe_b64(value: Any) -> bytes:
    if not isinstance(value, str):
        return b""
    try:
        return _vr._b64decode(value)
    except Exception:
        return b""


    # Asqav Compliance Receipt - ML-DSA-65 over canonical bytes (compliance or hash mode).
class AsqavNativeAdapter(FormatAdapter):

    name = "asqav-native"

    def detect(self, doc: dict) -> bool:
        if _is_hash_mode(doc):
            return True
        sig = doc.get("signature")
        # An ACTA receipt carries a lowercase-hex sig; decline it so the formats stay disjoint.
        if isinstance(sig, dict) and _is_lower_hex(sig.get("sig")):
            return False
        payload = doc.get("payload")
        if isinstance(payload, dict) and "previousReceiptHash" in payload:
            return True
        # A bare payload (the conformance-vector shape) is also Asqav-native.
        return "previousReceiptHash" in doc and "issuer_id" in doc

    def extract_signature(self, doc: dict) -> SignatureMaterial:
        if _is_hash_mode(doc):
            return SignatureMaterial(
                sig=_safe_b64(doc.get("signature_b64") or doc.get("signature", "")),
                alg=doc.get("algorithm", "ML-DSA-65"),
                kid=doc.get("key_id", ""),
            )
        env = _vr.normalise_envelope(doc)
        sig_obj = env.get("signature", {})
        if isinstance(sig_obj, str):
            sig_obj = {"alg": "ML-DSA-65", "kid": _payload(doc).get("issuer_id", ""), "sig": sig_obj}
        return SignatureMaterial(
            sig=_safe_b64(sig_obj.get("sig", "")),
            alg=sig_obj.get("alg", "ML-DSA-65"),
            kid=sig_obj.get("kid", ""),
        )

    def _signing_key_entry(self, doc: dict, jwks: dict) -> dict | None:
        """The one jwks entry this receipt's signature is checked against.

        Cloud receipts set kid to the issuer id; when that resolves nothing, the
        agent's own key answers, mirroring run(). agent_id is attacker-controlled,
        so the agent match trusts only a key whose published issuer_id equals the
        one the receipt claims inside its signed bytes.

        Every axis resolves through here. kid lives OUTSIDE the signed bytes, so a
        second independent lookup on it is attacker-steerable: a receipt could
        verify against a key found by the agent route while the revocation and
        issuer axes read a kid that resolves nothing and emit no axis at all. An
        axis that never exists cannot be blocked on, unlike one that reports
        SKIPPED, so the verdict would read PASS for a key the directory revoked.
        """
        payload = _payload(doc)
        return _vr.match_signing_key(
            jwks,
            self.extract_signature(doc).kid,
            payload.get("agent_id") or doc.get("agent_id"),
            payload.get("issuer_id"),
            payload.get("org_id") or doc.get("org_id"),
        )

    def resolve_key(self, doc: dict, key_provider: Any) -> tuple[bytes | None, str]:
        jwks = key_provider or {"keys": []}
        kid = self.extract_signature(doc).kid
        entry = self._signing_key_entry(doc, jwks)
        if entry is None:
            return None, f"kid {kid!r} not in jwks directory"
        pk = _vr._b64decode(entry["public_key"])
        status = entry.get("status")
        if kid and kid in (entry.get("issuer_id"), entry.get("kid")):
            return pk, f"resolved kid {kid} (status={status})"
        return pk, f"resolved agent key {entry.get('kid')} (status={status})"

    def signing_input(self, doc: dict) -> bytes:
        if _is_hash_mode(doc):
            return self._hash_mode_signing_input(doc)
        # Asqav signs the canonical bytes of the payload directly, no pre-hash.
        return asqav_jcs(_payload(doc))

    def _hash_mode_signing_input(self, doc: dict) -> bytes:
        """Rebuild the flat object the cloud's hash-mode path signs, then canonicalise.

        Mirrors ``agents.py::_build_signing_message`` hash-mode branch field-for-field;
        ``asqav_jcs`` sorts the keys, so insertion order is cosmetic but kept aligned.
        """
        flat = {
            "v": 1,
            "mode": "hash",
            "hash": doc.get("hash"),
            "hash_algo": doc.get("hash_algo") or "sha256",
            "metadata": doc.get("metadata") or {},
            "server_timestamp": doc.get("server_timestamp"),
            "action_id": doc.get("action_id"),
            "agent_id": doc.get("agent_id"),
            "org_id": doc.get("org_id"),
            "policy_digest": doc.get("policy_digest"),
            "policy_decision": doc.get("policy_decision"),
        }
        return asqav_jcs(flat)

    def chain_step(self, doc: dict) -> ChainStep:
        if _is_hash_mode(doc):
            # A hash-mode signature receipt carries no in-band chain link of its own.
            return ChainStep(prev_field=None, is_genesis=True, recompute=lambda _pred: "")
        prev = _payload(doc).get("previousReceiptHash")
        is_genesis = prev == _vr.FIRST_RECEIPT_SEED
        return ChainStep(
            prev_field=prev,
            is_genesis=is_genesis,
            recompute=lambda pred: sha256_hex(asqav_jcs(_payload(pred))),
        )

    def schema(self, doc: dict) -> tuple[str, str]:
        if _is_hash_mode(doc):
            missing = [f for f in _HASH_MODE_FIELDS if doc.get(f) is None and f != "policy_digest"]
            if missing:
                return "FAIL", f"hash-mode receipt missing fields: {','.join(missing)}"
            # A claim outside the signed field set is unauthenticated, whatever it
            # says, so refuse it rather than reporting on bytes nobody signed.
            unsigned = [f for f in _UNSIGNED_CLAIM_FIELDS if f in doc]
            if unsigned:
                return "FAIL", (
                    f"hash-mode receipt carries claim fields its signature does not "
                    f"cover: {','.join(unsigned)}"
                )
            return "PASS", "hash-mode signature receipt; required flat fields present"
        return _vr.check_structure(_payload(doc))

    def __init__(self) -> None:
        # Shared per instance, so a duplicate (issuer_id, nonce) pair is flagged (draft 5.7).
        self._seen_nonces: set[str] = set()

    def extra_axes(self, doc: dict, key_provider: Any) -> list[tuple[str, str, str]]:
        """Gate the verdict on expiry, the signing key's revocation status, and its issuer.

        A receipt signed by a revoked key, or by a key the directory publishes
        under a different issuer, must not PASS offline, matching the hosted
        /verify. No key resolved at all means the signature axis already reports no
        key, so those two axes only weigh in once a key is found. Both wire shapes name
        their issuer inside the signed bytes: issuer_id in compliance mode, org_id
        in hash mode.

        Resolution goes through the same entry the signature axis verifies
        against, so these axes weigh the key that actually signed rather than
        whatever the unsigned kid happens to name.
        """
        # Expiry reads only the signed bytes, so no key is needed. Hash mode signs no
        # expires_at, and reading the flat doc would gate on an uncovered field.
        hash_mode = _is_hash_mode(doc)
        signed = {} if hash_mode else _payload(doc)
        axes: list[tuple[str, str, str]] = [("expiry", *_vr.check_expiry(signed))]
        axes.append(("nonce", *_vr.check_nonce(signed, self._seen_nonces)))
        jwks = key_provider or {"keys": []}
        entry = self._signing_key_entry(doc, jwks)
        # Reported before the no-entry return, so a receipt binding no thumbprint
        # still says so rather than dropping the axis when nothing resolved.
        bound_alg, bound_pk = _resolved_key_material(entry)
        axes.append(("key_binding", *_vr.check_key_binding(signed, bound_alg, bound_pk)))
        # No database offline, so a claimed binding reports unresolved rather than
        # riding along as corroboration nobody checked
        axes.append(("counterparty", *_vr.check_counterparty_binding(signed)))
        axes.append(("payload_digest", *_vr.check_payload_digest(signed)))
        # Hash mode signs no issued_at, so skew reads the flat server_timestamp; without
        # this the oracle accepted a 2099 issue time the standalone verifier refuses.
        stamp = doc.get("server_timestamp", "") if hash_mode else signed.get("issued_at", "")
        axes.append(("skew", *_vr.check_skew(stamp)))
        if entry is None:
            return axes
        key_issuer = _vr.key_issuer_of(entry)
        # The resolved key is bound back to the issuer the signed bytes name.
        if _is_hash_mode(doc):
            issued_at = doc.get("server_timestamp", "")
            bind = _vr.check_org_binding(key_issuer, _vr.key_org_of(entry), doc.get("org_id"))
        else:
            payload = _payload(doc)
            issued_at = payload.get("issued_at", "")
            bind = _vr.check_issuer_binding(key_issuer, payload.get("issuer_id"))
        # Only a caller-verified anchor counts as trusted timing. This adapter pins no
        # TSA material, so it passes False: a forged anchor never rides a revoked key.
        res, note = _vr.check_key_status(
            entry.get("status"), issued_at, _vr.revoked_at_of(entry), False
        )
        return axes + [("key_status", res, note), ("issuer_bind", *bind)]

    def attestation(self, doc: dict) -> dict[str, Any]:
        """Surface the v:2 in-body ``signer``. None for v:1 and hash-mode.

        Read only from the signed payload, so a signer appended as loose
        metadata outside the canonical body is never surfaced (and would also
        break the signature).
        """
        if _is_hash_mode(doc):
            return {}
        signer = _payload(doc).get("signer")
        return {"signer": signer} if signer is not None else {}

    def keyed_digest(self, doc: dict) -> bool:
        """A hash-mode digest sealed with the org salt is keyed (criterion 438).

        hash_algo hmac-sha256 is internally consistent but not third-party
        re-derivable, so a fully-checked receipt reports verified_keyed, never
        plain verified. Read from the signed field set only.
        """
        return _is_hash_mode(doc) and doc.get("hash_algo") == "hmac-sha256"

File: typescript/src/verifier/adapters/asqavNative.ts
/**
 * Asqav-native adapter, a port of the Python oracle's `asqav_native.py`. Two wire shapes route here,
 * mutually exclusive at detection: the compliance envelope and the flat hash-mode receipt.
 */

import {
  FormatAdapter,
  type AxisCheck,
  type ChainStep,
  type ExtraAxis,
  type KeyProvider,
  type SignatureMaterial,
} from "../adapter.js";
import {
  JCS_UTF16_CUTOVER,
  asqavJcs,
  asqavJcsPreCutover,
  hasSupplementaryMemberName,
} from "../canonical.js";
import { sha256Hex } from "../crypto.js";
import { isLowerHex } from "./acta.js";
import {
  FIRST_RECEIPT_SEED,
  b64decode,
  checkExpiry,
  checkIssuerBinding,
  checkCounterpartyBinding,
  checkKeyBinding,
  checkKeyStatus,
  checkPayloadDigest,
  checkSkew,
  checkNonce,
  checkOrgBinding,
  checkStructure,
  keyIssuerOf,
  keyOrgOf,
  matchSigningKey,
  normaliseEnvelope,
  revokedAtOf,
} from "../vrShim.js";

/** Field set the cloud's hash-mode signer canonicalises (for the structure check). */
const HASH_MODE_FIELDS = [
  "v",
  "mode",
  "hash",
  "hash_algo",
  "metadata",
  "server_timestamp",
  "action_id",
  "agent_id",
  "org_id",
  "policy_digest",
  "policy_decision",
] as const;

function isRecord(v: unknown): v is Record<string, unknown> {
  return v !== null && typeof v === "object" && !Array.isArray(v);
}

/** True for a flat hash-mode signature receipt (mode=hash, null payload, a sig). */
// Claims belonging to the signed payload; hash mode signs the flat fields only,
// so a thumbprint or seq pasted onto one binds nothing.
const UNSIGNED_CLAIM_FIELDS = ["issuer_id", "previousReceiptHash", "key_thumbprint", "seq"];

/**
 * alg plus raw public-key bytes of the resolved directory entry. The directory publishes standard
 * base64 while an AKP thumbprint is over unpadded base64url, so the decode precedes the digest.
 */
function resolvedKeyMaterial(
  entry: Record<string, unknown> | null,
): readonly [unknown, Uint8Array | null] {
  if (entry === null) return [null, null];
  try {
    return [entry.alg, b64decode(entry.public_key as string)];
  } catch {
    return [entry.alg, null];
  }
}

function isHashMode(doc: Record<string, unknown>): boolean {
  // Routing reads the shape of the signed unit only: a record payload is the
  // compliance signed unit, any other shape routes here (mirrors _is_hash_mode).
  if (doc.mode !== "hash" || isRecord(doc.payload)) return false;
  return Boolean(doc.signature_b64 || doc.signature);
}

/** Normalise to the signed payload regardless of envelope nesting. */
function payloadOf(doc: Record<string, unknown>): Record<string, unknown> {
  const env = normaliseEnvelope(doc);
  const p = env.payload;
  return isRecord(p) ? p : env;
}

/** Decode signature material; empty on malformed input so verify FAILs, never crashes. */
function safeB64(value: unknown): Uint8Array {
  if (typeof value !== "string") return new Uint8Array(0);
  try {
    return b64decode(value);
  } catch {
    return new Uint8Array(0);
  }
}

export class AsqavNativeAdapter extends FormatAdapter {
  readonly name = "asqav-native";

  /** Shared per instance, so a duplicate (issuer_id, nonce) pair is flagged (draft 5.7). */
  private readonly seenNonces = new Set<string>();

  detect(doc: Record<string, unknown>): boolean {
    if (isHashMode(doc)) return true;
    const sig = doc.signature;
    // An ACTA receipt carries a lowercase-hex sig; decline it so the formats stay disjoint.
    if (isRecord(sig) && isLowerHex(sig.sig)) return false;
    const payload = doc.payload;
    if (isRecord(payload) && "previousReceiptHash" in payload) return true;
    // A bare payload (the conformance-vector shape) is also Asqav-native.
    return "previousReceiptHash" in doc && "issuer_id" in doc;
  }

  extractSignature(doc: Record<string, unknown>): SignatureMaterial {
    if (isHashMode(doc)) {
      return {
        sig: safeB64(doc.signature_b64 || doc.signature || ""),
        alg: (doc.algorithm as string) ?? "ML-DSA-65",
        kid: (doc.key_id as string) ?? "",
      };
    }
    const env = normaliseEnvelope(doc);
    let sigObj = env.signature;
    if (typeof sigObj === "string") {
      sigObj = { alg: "ML-DSA-65", kid: payloadOf(doc).issuer_id ?? "", sig: sigObj };
    }
    const so = isRecord(sigObj) ? sigObj : {};
    return {
      sig: safeB64(so.sig ?? ""),
      alg: (so.alg as string) ?? "ML-DSA-65",
      kid: (so.kid as string) ?? "",
    };
  }

  /**
   * The one JWKS entry this receipt's signature is checked against. Every axis resolves through here: kid
   * sits outside the signed bytes, so a second independent lookup would be attacker-steerable.
   */
  private signingKeyEntry(
    doc: Record<string, unknown>,
    jwks: Record<string, unknown>,
  ): Record<string, unknown> | null {
    const payload = payloadOf(doc);
    return matchSigningKey(
      jwks,
      this.extractSignature(doc).kid,
      payload.agent_id ?? doc.agent_id,
      payload.issuer_id,
      payload.org_id ?? doc.org_id,
    );
  }

  resolveKey(
    doc: Record<string, unknown>,
    keyProvider: KeyProvider,
  ): readonly [Uint8Array | null, string] {
    const jwks = (keyProvider ?? { keys: [] }) as Record<string, unknown>;
    const kid = this.extractSignature(doc).kid;
    const entry = this.signingKeyEntry(doc, jwks);
    if (entry === null) {
      return [null, `kid '${kid}' not in jwks directory`];
    }
    const pk = b64decode(entry.public_key as string);
    const status = (entry.status as string) ?? null;
    if (kid && (kid === entry.issuer_id || kid === entry.kid)) {
      return [pk, `resolved kid ${kid} (status=${status})`];
    }
    return [pk, `resolved agent key ${entry.kid} (status=${status})`];
  }

  signingInput(doc: Record<string, unknown>): Uint8Array {
    if (isHashMode(doc)) {
      return this.hashModeSigningInput(doc);
    }
    // Asqav signs the canonical bytes of the payload directly, no pre-hash.
    return asqavJcs(payloadOf(doc));
  }

  // Only a payload-mode receipt issued before the cutover with a member name above U+FFFF has
  // a dated dialect; hash-mode members are ASCII, so both orders coincide there.
  preCutoverSigningInput(doc: Record<string, unknown>): Uint8Array | null {
    if (isHashMode(doc)) return null;
    const payload = payloadOf(doc);
    if (!hasSupplementaryMemberName(payload)) return null;
    const issued = Date.parse(String(payload.issued_at ?? ""));
    if (Number.isNaN(issued) || issued >= Date.parse(JCS_UTF16_CUTOVER)) return null;
    return asqavJcsPreCutover(payload);
  }

  private hashModeSigningInput(doc: Record<string, unknown>): Uint8Array {
    const flat = {
      v: 1,
      mode: "hash",
      hash: doc.hash ?? null,
      hash_algo: doc.hash_algo ?? "sha256",
      metadata: doc.metadata ?? {},
      server_timestamp: doc.server_timestamp ?? null,
      action_id: doc.action_id ?? null,
      agent_id: doc.agent_id ?? null,
      org_id: doc.org_id ?? null,
      policy_digest: doc.policy_digest ?? null,
      policy_decision: doc.policy_decision ?? null,
    };
    return asqavJcs(flat);
  }

  chainStep(doc: Record<string, unknown>): ChainStep {
    if (isHashMode(doc)) {
      // A hash-mode signature receipt carries no in-band chain link of its own.
      return { prevField: null, isGenesis: true, recompute: () => "" };
    }
    const prev = payloadOf(doc).previousReceiptHash;
    const isGenesis = prev === FIRST_RECEIPT_SEED;
    return {
      prevField: (prev as string | null) ?? null,
      isGenesis,
      recompute: (pred) => sha256Hex(asqavJcs(payloadOf(pred))),
    };
  }

  schema(doc: Record<string, unknown>): AxisCheck {
    if (isHashMode(doc)) {
      const missing = HASH_MODE_FIELDS.filter(
        (f) => (doc[f] === undefined || doc[f] === null) && f !== "policy_digest",
      );
      if (missing.length > 0) {
        return ["FAIL", `hash-mode receipt missing fields: ${missing.join(",")}`];
      }
      // A claim outside the signed field set is unauthenticated whatever it says.
      const unsigned = UNSIGNED_CLAIM_FIELDS.filter((f) => f in doc);
      if (unsigned.length > 0) {
        return [
          "FAIL",
          `hash-mode receipt carries claim fields its signature does not cover: ${unsigned.join(",")}`,
        ];
      }
      return ["PASS", "hash-mode signature receipt; required flat fields present"];
    }
    return checkStructure(payloadOf(doc));
  }

  // Gate on expiry, signing key revocation status, and its issuer (mirrors Python extra_axes).
  // The key axes are a no-op when the key is absent; the signature axis handles that.
  extraAxes(doc: Record<string, unknown>, keyProvider: KeyProvider): ExtraAxis[] {
    const hashMode = isHashMode(doc);
    // Expiry reads only the signed bytes, so no key is needed. Hash mode signs no
    // expires_at, and reading the flat doc would gate on an uncovered field.
    const axes: ExtraAxis[] = [["expiry", ...checkExpiry(hashMode ? {} : payloadOf(doc))]];
    axes.push(["nonce", ...checkNonce(hashMode ? {} : payloadOf(doc), this.seenNonces)]);
    const jwks = (keyProvider ?? { keys: [] }) as Record<string, unknown>;
    const entry = this.signingKeyEntry(doc, jwks);
    // Reported before the no-entry return, so a receipt binding no thumbprint
    // still says so rather than dropping the axis when nothing resolved.
    const [boundAlg, boundPk] = resolvedKeyMaterial(entry);
    const signedUnit = hashMode ? {} : payloadOf(doc);
    axes.push(["key_binding", ...checkKeyBinding(signedUnit, boundAlg, boundPk)]);
    // No database offline, so a claimed binding reports unresolved rather than
    // riding along as corroboration nobody checked
    axes.push(["counterparty", ...checkCounterpartyBinding(signedUnit)]);
    axes.push(["payload_digest", ...checkPayloadDigest(signedUnit)]);
    // Hash mode signs no issued_at, so skew reads the flat server_timestamp there
    const stamp = hashMode ? doc.server_timestamp : signedUnit.issued_at;
    axes.push(["skew", ...checkSkew(stamp)]);
    if (entry === null) return axes;
    // Both wire shapes name their issuer inside the signed bytes: issuer_id in
    // compliance mode, org_id in hash mode.
    const issuedAt = hashMode
      ? String(doc.server_timestamp ?? "")
      : String(payloadOf(doc).issued_at ?? "");
    // Offline anchor presence is unverifiable (anchors unsigned); pass false
    // so a forged anchor never rides a revoked key to PASS.
    const status = (entry.status as string) ?? null;
    const [res, note] = checkKeyStatus(status, issuedAt, revokedAtOf(entry), false);
    // The key that verified is bound back to the issuer the signed bytes name.
    const keyIssuer = keyIssuerOf(entry);
    const [bindRes, bindNote] = hashMode
      ? checkOrgBinding(keyIssuer, keyOrgOf(entry), doc.org_id)
      : checkIssuerBinding(keyIssuer, payloadOf(doc).issuer_id);
    return [
      ...axes,
      ["key_status", res, note],
      ["issuer_bind", bindRes, bindNote],
    ];
  }

  // Hash mode signs the flat field set only, so a seq sitting there binds nothing.
  seqOf(doc: Record<string, unknown>): unknown {
    if (isHashMode(doc)) return null;
    return payloadOf(doc).seq ?? null;
  }

  // Surface the v:2 in-body signer. null for v:1 and hash-mode. Read only from
  // the signed payload, so a signer appended as loose metadata is never surfaced.
  attestation(doc: Record<string, unknown>): Record<string, unknown> {
    if (isHashMode(doc)) return {};
    const signer = payloadOf(doc).signer;
    return signer !== undefined && signer !== null ? { signer } : {};
  }

  /**
   * A hash-mode digest sealed with the org salt is keyed (criterion 438): internally consistent but not
   * third-party re-derivable, so a fully-checked receipt reports verified_keyed, never plain verified.
   */
  keyedDigest(doc: Record<string, unknown>): boolean {
    return isHashMode(doc) && doc.hash_algo === "hmac-sha256";
  }
}
